smack-experimental/src/main/java/org/jivesoftware/smackx/message_markup/MessageMarkup.java

```diff
@@ -0,0 +1,220 @@
+            }
+        }
+
+        quotes.add(new BlockQuoteElement(start, end));
+        return this;
+    }
+
+    /**
+     * Mark a section of a message as a code block.
+     *
+     * @param start start index
+     * @param end end index
+     * @return builder
+     */
+    public MessageMarkup setCodeBlock(int start, int end) {
+        verifyStartEnd(start, end);
+
+        codes.add(new CodeBlockElement(start, end));
+        return this;
+    }
+
+    /**
+     * Begin a list.
+     *
+     * @return list builder
+     */
+    public MessageMarkupListBuilder beginList() {
+        return new MessageMarkupListBuilder(this);
+    }
+
+    public static final class MessageMarkupListBuilder {
+        private final MessageMarkup markup;
+        private final ArrayList<ListElement.ListEntryElement> entries = new ArrayList<>();
+        private int end = -1;
+
+        private MessageMarkupListBuilder(MessageMarkup markup) {
+            this.markup = markup;
+        }
+
+        /**
+         * Add an entry to the list.
+         * The start index of an entry must correspond to the end index of the previous entry
+         * (if a previous entry exists.)
+         *
+         * @param start start index
+         * @param end end index
+         * @return list builder
+         */
+        public MessageMarkupListBuilder addEntry(int start, int end) {
+            verifyStartEnd(start, end);
+
+            ListElement.ListEntryElement last = entries.size() == 0 ? null : entries.get(entries.size() - 1);
+            if (last != null && start != this.end) {
+                throw new IllegalArgumentException("Next entries start must be equal to last entries end.");
+            }
+            entries.add(new ListElement.ListEntryElement(start));
+            this.end = end;
+
+            return this;
+        }
+
+        /**
+         * End the list.
+         *
+         * @return builder
+         */
+        public MessageMarkup endList() {
+            if (entries.size() > 0) {
+                ListElement.ListEntryElement first = entries.get(0);
+                ListElement list = new ListElement(first.getStart(), end, entries);
+                markup.lists.add(list);
+            }
+
+            return markup;
+        }
+    }
+
+    /**
+     * Build a Message Markup element.
+     *
+     * @return extension element
+     */
+    public MarkupElement build() {
+        List<MarkupElement.MarkupChildElement> children = new ArrayList<>();
+        children.addAll(spans);
+        children.addAll(quotes);
+        children.addAll(codes);
+        children.addAll(lists);
+        return new MarkupElement(children);
+    }
+
+    private static void verifyStartEnd(int start, int end) {
+        if (start >= end || start < 0) {
+            throw new IllegalArgumentException("Start value MUST be greater equal than 0 and MUST be smaller than end value.");
```

Comment:
It is always a good idea to also return the actual values which where verified in the message that you are returning, be it via an exception or the log subsystem.